File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int left = 0;
        int right = 0;
        int maxFreq = 0;
        int maxLen = 0;
        unordered_map<char, int> mpp;
        while(right < s.size()){
            mpp[s[right]]++;
            maxFreq = max(maxFreq, mpp[s[right]]);
            int length = right - left + 1;
            if(length - maxFreq > k){
                mpp[s[left]]--;
                left++;
            }
            maxLen = max(maxLen, right - left + 1);
            right++;
        }
        return maxLen;
    }
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] freq = new int[26];

        int maxFreq = 0;
        int left = 0;
        int maxLength = 0;

        for(int right = 0; right < s.length(); right++){
            int index = s.charAt(right) - 'A';
            freq[index]++;

            if(freq[index] > maxFreq){
                maxFreq = freq[index];
            }

            while((right - left + 1) - maxFreq > k){
                freq[s.charAt(left) - 'A']--;
                left++;
            }

            int windowSize = right - left + 1;

            if(windowSize > maxLength){
                maxLength = windowSize;
            }
        }
        return maxLength;
    }
}